File: package/umt_python/src/ip/subnet_mask_to_cidr.py
def subnet_mask_to_cidr(subnet_mask: str) -> int:
    """
    Converts a subnet mask to CIDR notation.

    Args:
        subnet_mask: IPv4 subnet mask (e.g., "255.255.255.0")

    Returns:
        CIDR notation (0-32)

    Raises:
        ValueError: If subnet mask is invalid

    Example:
        >>> subnet_mask_to_cidr("255.255.255.0")
        24
        >>> subnet_mask_to_cidr("255.255.0.0")
        16
    """
    if not subnet_mask:
        raise ValueError("Subnet mask is required")

    octets = subnet_mask.split(".")
    if len(octets) != 4:
        raise ValueError("Invalid subnet mask format")

    try:
        nums = [int(octet) for octet in octets]
    except ValueError:
        raise ValueError("Invalid subnet mask format") from None

    mask = 0
    for num in nums:
        if num < 0 or num > 255:
            raise ValueError("Invalid subnet mask format")
        mask = (mask << 8) | num

    # Host bits of a valid prefix mask are 0 or 2^n-1, so host & (host + 1)
    # is 0. 0xFFFFFFFF ^ mask is a 32-bit invert (Python ints are unbounded).
    host = 0xFFFFFFFF ^ mask
    if host & (host + 1) != 0:
        raise ValueError("Invalid subnet mask: must be consecutive 1s followed by 0s")

    return mask.bit_count()

File: package/umt_rust/src/ip/subnet_mask_to_cidr.rs
/// Converts a subnet mask to CIDR notation
///
/// # Arguments
/// * `subnet_mask` - IPv4 subnet mask (e.g., "255.255.255.0")
///
/// # Returns
/// * `Ok(u8)` - CIDR notation (0-32)
/// * `Err` - If subnet mask is invalid
///
/// # Examples
/// ```
/// use umt_rust::ip::subnet_mask_to_cidr;
/// assert_eq!(subnet_mask_to_cidr("255.255.255.0").unwrap(), 24);
/// assert_eq!(subnet_mask_to_cidr("255.255.0.0").unwrap(), 16);
/// ```
pub fn subnet_mask_to_cidr(subnet_mask: &str) -> Result<u8, String> {
    if subnet_mask.is_empty() {
        return Err("Subnet mask is required".to_string());
    }

    let mut mask = 0u32;
    let mut octet_count = 0u8;

    for octet in subnet_mask.split('.') {
        if octet_count == 4 {
            return Err("Invalid subnet mask format".to_string());
        }

        let number: u8 = octet
            .parse()
            .map_err(|_| "Invalid subnet mask format".to_string())?;

        mask = (mask << 8) | u32::from(number);
        octet_count += 1;
    }

    if octet_count != 4 {
        return Err("Invalid subnet mask format".to_string());
    }

    let host = !mask;
    if host & host.wrapping_add(1) != 0 {
        return Err("Invalid subnet mask: must be consecutive 1s followed by 0s".to_string());
    }

    Ok(mask.count_ones() as u8)
}
